File: convert/example1_todo/java/todo/src/test/java/dev/practice/todo/api/service/TodoServiceTest.java
package dev.practice.todo.api.service;

import dev.practice.todo.domain.Todo;
import dev.practice.todo.domain.TodoRepository;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.test.context.junit.jupiter.SpringExtension;

import java.time.LocalDateTime;
import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.BDDMockito.given;

@ExtendWith(SpringExtension.class) // 프로젝트에서 개발한 애들을 띄우지는 않고.. 스프링 기본 컨테이너를 띄우는듯, (service 직접 생성함)
public class TodoServiceTest {

    @MockBean
    private TodoRepository repository;

    private TodoService service;

    private Todo stub;

    @BeforeEach
    public void setUp() {
        service = new TodoService(repository);
        stub = todoStub();
    }

    @Test
    public void 한개의_TODO를_반환해야한다() {
        // Given
        given(repository.findById(1L)).willReturn(Optional.of(stub));

        // When
        Todo actual = service.getTodo(1L);

        // Then
        assertThat(actual).isNotNull();
        assertThat(actual).isEqualTo(stub);
    }


    public Todo todoStub() {
        return Todo.builder()
                .id(1L)
                .title("테스트")
                .description("테스트 상세")
                .done(false)
                .createdAt(LocalDateTime.now())
                .updatedAt(LocalDateTime.now()).build();
    }


}

File: convert/example1_todo/kotlin/todo/src/test/java/dev/practice/todo/api/service/TodoServiceTest.kt
package dev.practice.todo.api.service

import com.ninjasquad.springmockk.MockkBean
import dev.practice.todo.domain.Todo
import dev.practice.todo.domain.TodoRepository
import io.mockk.every
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.data.repository.findByIdOrNull
import org.springframework.test.context.junit.jupiter.SpringExtension
import java.time.LocalDateTime

@ExtendWith(SpringExtension::class)
class TodoServiceTests {

    @MockkBean
    lateinit var repository: TodoRepository

    lateinit var service: TodoService

    val stub: Todo by lazy {
        Todo(
            id = 1,
            title = "테스트",
            description = "테스트 상세",
            done = false,
            createdAt = LocalDateTime.now(),
            updatedAt = LocalDateTime.now(),
        )
    }

    @BeforeEach
    fun setUp() {
        service = TodoService(repository)
    }


    @Test
    fun `한개의 TODO를 반환해야한다`() {
        // Given
        every { repository.findByIdOrNull(1L) } returns stub

        // When
        val actual = service.getTodo(1L)

        // Then
        assertThat(actual).isNotNull
        assertThat(actual).isEqualTo(stub)
    }

}